### PR TITLE
2D U-Net Segmentation of Prostate MRI Slices

### PR DESCRIPTION
**Pull request by ryanmaher-student**
This pull request contains my implementation of a 2D U-Net applied to the HipMRI Study on Prostate Cancer. It contains modules.py, train.py, predict.py and dataset.py (and utils.py containing a helper function for the provided loading function). For more infomation see the associated README, which provides a more in depth explanation of the algorithm, and a description of each of the components and their purposes.

Thanks, Ryan Maher - 47484212

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**  
1. **Solves problem**: None of the labels achieved the 0.75 DICE score. The training loss plot was not provided. As mentioned in the readme and [train.py](https://github.com/ryanmaher-student/PatternAnalysis-2024/blob/topic-recognition/recognition/Prostate-2D-Segmentation-47484212/train.py#L23) script, the model has only trained 3 epochs and has not converged yet. The figures in the readme failed to display as no image was uploaded.  
2. **Implementation**: In the [training](https://github.com/ryanmaher-student/PatternAnalysis-2024/blob/topic-recognition/recognition/Prostate-2D-Segmentation-47484212/train.py#L22C1-L25C1) script, hyperparameters were hard-coded instead of passing through ArgParse. In the [predicting](https://github.com/ryanmaher-student/PatternAnalysis-2024/blob/topic-recognition/recognition/Prostate-2D-Segmentation-47484212/predict.py#L21) script, DICE list was named of classloss which in unreasonable. Otherwise, all good.  
3. **Design**: well-designed.  
4. **Comment**: sufficient comments.
5. **Difficulty**: **Easy Difficulty** is capped at 10/20 marks.

**Comment by ryanmaher-student** [comment]
> **This is an initial inspection, no action is required at this point**
> 
> 1. **Solves problem**: None of the labels achieved the 0.75 DICE score. The training loss plot was not provided. As mentioned in the readme and [train.py](https://github.com/ryanmaher-student/PatternAnalysis-2024/blob/topic-recognition/recognition/Prostate-2D-Segmentation-47484212/train.py#L23) script, the model has only trained 3 epochs and has not converged yet. The figures in the readme failed to display as no image was uploaded.
> 2. **Implementation**: In the [training](https://github.com/ryanmaher-student/PatternAnalysis-2024/blob/topic-recognition/recognition/Prostate-2D-Segmentation-47484212/train.py#L22C1-L25C1) script, hyperparameters were hard-coded instead of passing through ArgParse. In the [predicting](https://github.com/ryanmaher-student/PatternAnalysis-2024/blob/topic-recognition/recognition/Prostate-2D-Segmentation-47484212/predict.py#L21) script, DICE list was named of classloss which in unreasonable. Otherwise, all good.
> 3. **Design**: well-designed.
> 4. **Comment**: sufficient comments.
> 5. **Difficulty**: **Easy Difficulty** is capped at 10/20 marks.

Hi, I have uploaded the missing images for the pdf and pushed changes to the pdf and predict.py file to show the Dice Coefficient instead of the Dice loss which I was displaying.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Model training not carried out.</td><td>-3</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. No train graphs, results.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Pull Request has problems: Late submission.</td><td>-2</td></tr>
	<tr><td></td><td>No Feedback required. 1</td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-19</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by ryanmaher-student** [comment]
Hi, I am curious what the criteria is for model training being carried out was as I thought we were not meant to upload our trained model? my initial pull request sufficiently trained the model to above the required metrics. Also, is the late submission penalty relating to the date of my initial pull request? Just curious as that was before the date of my approved extension. If this was relating to my commits following the intial inspection, these can be ignored if they reduce my mark as they were just to clarify that my model had converged and was acheiving the required scores.

**Comment by shakes76** [comment]
Approved extension +2

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally
